### PR TITLE
add line number and line event support

### PR DESCRIPTION
**Pull request by HParker**
Now we can emit line events and line numbers like Ruby does!

example, 

```
1 +
  2 +
  3
```

```
[1,
  :RUBY_EVENT_LINE,
  [:putobject, 1],
  2,
  [:putobject, 2],
  1,
  [:send, {:mid=>:+, :flag=>16, :orig_argc=>1}, nil],
  3,
  [:putobject, 3],
  2,
  [:send, {:mid=>:+, :flag=>16, :orig_argc=>1}, nil],
  [:leave]]]
```

Related: https://github.com/Shopify/yarp/discussions/667

### CODE REVIEW

**Review by jemmaissroff — CHANGES_REQUESTED** [review]
Thanks for working on this! Left a couple qs

File: ext/yarp/extension.c [hunk 1]
```diff
@@ -428,11 +428,31 @@ compile(VALUE self, VALUE string) {
   yp_parser_init(&parser, RSTRING_PTR(string), length, NULL);
 
   yp_node_t *node = yp_parse(&parser);
-  VALUE result = yp_compile(node);
+
+  // TODO: this hack
```
[the remark this answers is not in the document]
Comment (HParker) [reply ext/yarp/extension.c:432]: I should have deleted this. I initially just `malloc`ed this to size 100, but forgot to remove the comment. I think the `realloc` way isn't that hacky, though I wish I didn't have to store it this way.

File: ext/yarp/extension.c [hunk 2]
```diff
@@ -428,11 +428,31 @@ compile(VALUE self, VALUE string) {
...
+  int index = 0;
+  int size = 10;
+  char ** newline_locations = malloc(sizeof(char*) * size);
+
+  char *source = StringValueCStr(string);
+  for (int i = 0; i < strlen(source); i++) {
```
[the remark this answers is not in the document]
Comment (HParker) [reply ext/yarp/extension.c:438]: Ah! it does warn! I wonder if we can turn those warnings into CI failures.

[the remark this answers is not in the document]
Comment (jemmaissroff) [reply ext/yarp/extension.c]: Good call, I'll look into it!

**Comment by enebo** [comment]
> @enebo thanks for the thoughts!
> 
> > but I think you will emit some line numbers differently
> 
> Can you say more here? Which line numbers do you think will be different?

I think maybe it is possible for something containing a statement but is not a statement to emit the line number before where it would have naturally been emitted.  Or not :) I just see it as a different mechanism for the same thing but it is unclear to me it will behave exactly the same.  Figuring out this is complicated if I look at module (which contains stmts) I see it is 'primary' if I burrow down the grammar I see stmts -> stmt -> expr -> then I get lost so I cannot tell if module would be a stmt or not.   If not then you would be putting a line before it would be with the current parser. 

 This would be where comparing iseqs could probably show it one way or another.  If compatibility is paramount to merging with MRI then I think you will need to do this anyways (unless compile is comparing iseqs already).

> > whether you should be comparing every node/insn to see if jumped to a new line?
> 
> And then marking the node?

I think if the line number check is only when a node is marked it would be less work but only because I think many nodes are not marked newline.  It is trading looking for any node which changes the start line to just ones with something on a struct.  From my perspective I see serialize and compile having to replicate the same logic even if this does work.  So it may be more convenient during parse than during compile+serialize.  With that said adding data to a struct has its own tradeoffs.

The other issue (and this perhaps should be in a discussion is having to binary search to determine line will have some impact over potentially storing the line while the parsing is originally occurring.  I don't want to complicate things though.

**Comment by jemmaissroff** [comment]
Thanks for the thoughts @enebo. We can look into the serialization case. What serialization API for newlines would make most sense to you? We're wary of attaching newline information to every node struct because this would bloat the nodes unnecessarily. I wonder if attaching the same `newline_locations` list would make sense instead?

Then we could pull that list calculation out of compile, and re-use it for serialization, so the logic wouldn't be duplicated

**Comment by HParker** [comment]
I am still trying to understand what the tree would look like with newline nodes attached. is the idea,

"any node that changes the line number should include a newline child node"

a dump => load of the example in the description gives,

```
ProgramNode(0...13)(                                  
  ScopeNode(0...0)([]),                               
  StatementsNode(0...13)(                             
    [CallNode(0...13)(                                
       CallNode(0...7)(IntegerNode(0...1)(), nil, PLUS(2...3)("+"), nil, ArgumentsNode(6...7)([IntegerNode(6...7)()]), nil, nil, "+"),
       nil,                                           
       PLUS(8...9)("+"),                              
       nil,                                           
       ArgumentsNode(12...13)([IntegerNode(12...13)()]),
       nil,
       nil,
       "+"
     )]
  )
)
```

Would a tree including line numbers look like,

```
ProgramNode(0...13)(                                  
  ScopeNode(0...0)([]),                               
  StatementsNode(0...13)(                             
    [CallNode(0...13)(                                
       CallNode(0...7)(IntegerNode(0...1)(<LINE 1 NODE>), nil, PLUS(2...3)("+"), nil, ArgumentsNode(6...7)([<LINE 2 NODE>, IntegerNode(6...7)()]), nil, nil, "+"),
       nil,                                           
       PLUS(8...9)("+"),                              
       nil,                                           
       ArgumentsNode(12...13)([<LINE 3 NODE>, IntegerNode(12...13)()]),
       nil,
       nil,
       "+"
     )]
  )
)
```

I don't think this is right.

I am still struggling to understand where it would make sense to include line number nodes if we are only emitting them when the line number changes since execution can go forwards and backwards across line numbers.

**Comment by HParker** [comment]
Alternatively,

would runtimes have all the line information they need if `location` nodes include a line number?

**Comment by HParker** [comment]
Happy to split it up, but since the conversation is happening here, I added support for line:col information on location nodes in serialize. (and also broke all the tests, working on it) I think this is a potential way to do newline information in serialize without adding anything to the node.

```
ProgramNode(0...13)1:0-3:3(
  ScopeNode(0...0)1:0-1:0([]),
  StatementsNode(0...13)1:0-3:3(
    [CallNode(0...13)1:0-3:3(
       CallNode(0...7)1:0-2:4(IntegerNode(0...1)1:0-1:1(), nil, PLUS(2...3)1:2-2:0("+"), nil, ArgumentsNode(6...7)2:3-2:4([IntegerNode(6...7)2:3-2:4()]), nil, nil, "+"),
       nil,
       PLUS(8...9)2:5-3:0("+"),
       nil,
       ArgumentsNode(12...13)3:3-3:3([IntegerNode(12...13)3:3-3:3()]),
       nil,
       nil,
       "+"
     )]
  )
)
```

**Comment by enebo** [comment]
@HParker I will update the description in #667 with a more comprehensive description since we uncovered more details since I originally made it.

**Comment by enebo** [comment]
@HParker ah.  I did not realize parse errors already didn't know line number.  That is an excellent point.

**Comment by eregon** [comment]
I wrote above:

> Maybe it's fine to limit max line and column to 2^16 (64k), in which case it's the same as two ints. But still likely more memory than varints for offsets. It would also mean the need to have an array of line number to byte offset to be able to extract the relevant substring in the source. So that's one extra indirection for reading from the source. OTOH line&col->offset is O(1) with that table vs O(log n) for the binary search for offset->line&col.

Actually that's not true, while line->offset is O(1) with a table of line to offset, it's not that simple for column, since column is expressed in number of codepoints, not number of bytes. And we can't just have a table of column to offset, that's way too big. So if we store one of the two, offset seems better overall (incidentally, already what Truffle stores and it does the binary search to get line & column info from it).

For Coverage and :line TracePoint, I think we could just compute these line numbers lazily, only when Coverage or :line TracePoint (actually can even compute it lazily on TracePoint#lineno) is used. Then we don't need anything but offsets in the serialized form.

**Comment by HParker** [comment]
I am going to close this since I don't think it will merge in its current state, but I found the conversation super helpful and I think we can get closer in another attempt.

**Pull request closed without merging**